streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2472,18 +2472,58 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
     }
 
   /**
-   * Interrupts the stream if it does not produce a value after d duration.
+   * Ends the stream if it does not produce a value after d duration.
    */
   final def timeout(d: Duration): ZStream[R with Clock, E, O] =
     ZStream[R with Clock, E, O] {
       self.process.map { next =>
         next.timeout(d).flatMap {
-          case Some(a) => ZIO.succeedNow(a)
-          case None    => ZIO.interrupt
+          case Some(a) => Pull.emit(a)
+          case None    => Pull.end
         }
       }
     }
 
+  /**
+   * Ends the stream with given error if it does not produce a value after d duration.
+   */
+  final def timeoutErr[E1 >: E](e: E1)(d: Duration): ZStream[R with Clock, E1, O] =
```

Comment:
How about timeoutError and timeoutErrorCause?

Reply:
So that `timeoutErrorCause` uses `ZIO.die`?

Reply:
@iravid done, working on the other suggestions